File: src/ovi_voice_assistant/device_manager.py
"""Manages connections to all configured Voice PE devices."""

import asyncio
import logging
from dataclasses import dataclass

from ovi_voice_assistant.codec import create_codec
from ovi_voice_assistant.config import DeviceConfig, Settings
from ovi_voice_assistant.device_connection import DeviceConnection
from ovi_voice_assistant.music import MusicGroup
from ovi_voice_assistant.scheduler import Scheduler
from ovi_voice_assistant.transport.wifi import WiFiTransport
from ovi_voice_assistant.voice_assistant import VoiceAssistant

logger = logging.getLogger(__name__)

# How long to wait for competing wake events before picking a winner.
ARBITRATION_WINDOW_S = 0.5


@dataclass
class _WakeCandidate:
    connection: DeviceConnection
    score: int
    wake_word: str


class DeviceManager:
    """Creates and manages DeviceConnection instances for WiFi devices."""

    def __init__(
        self,
        devices: list[DeviceConfig],
        settings: Settings,
        pipeline: VoiceAssistant,
        tts_rate: int,
    ) -> None:
        self._connections: list[DeviceConnection] = []
        self._multi_device = len(devices) > 1

        # Wake-word arbitration state
        self._wake_candidates: list[_WakeCandidate] = []
        self._arbitration_handle: asyncio.TimerHandle | None = None

        # Shared music group for synchronized multi-device playback.
        # Created here so all devices share the same player/queue.
        music_codec = create_codec(settings.transport.codec, 48000, channels=2)
        self._music_group = MusicGroup(
            sample_rate=music_codec.sample_rate, channels=music_codec.channels
        )

        # Only use arbitration callback when there are multiple devices.
        on_wake = self._on_wake if self._multi_device else None

        for device in devices:
            transport = WiFiTransport(device.host, device.port, device.encryption_key)
            codec = create_codec(settings.transport.codec, tts_rate)
            self._connections.append(
                DeviceConnection(
                    transport,
                    codec,
                    pipeline,
                    settings,
                    on_wake=on_wake,
                    name=device.host,
                    music_group=self._music_group,
                )
            )

    async def start(self) -> None:
        for conn in self._connections:
            await conn.start()
        if self._multi_device:
            logger.info(
                "Managing %d device(s) — multi-device arbitration enabled (window=%.1fs)",
                len(self._connections),
                ARBITRATION_WINDOW_S,
            )
        else:
            logger.info("Managing 1 device")

    async def stop(self) -> None:
        if self._arbitration_handle is not None:
            self._arbitration_handle.cancel()
        await self._music_group.stop()
        for conn in self._connections:
            await conn.stop()

    async def announce_all(self, text: str) -> None:
        """Announce text on every connected device."""
        for conn in self._connections:
            conn.announce(text)

    def set_scheduler(self, scheduler: Scheduler) -> None:
        """Attach the scheduler to all device connections."""
        for conn in self._connections:
            conn.set_scheduler(scheduler)

    def set_memory(self, memory) -> None:
        """Attach Hindsight memory to all device connections."""
        for conn in self._connections:
            conn.set_memory(memory)

    # -- Wake-word arbitration --

    async def _on_wake(
        self, connection: DeviceConnection, score: int, wake_word: str
    ) -> None:
        """Called by any DeviceConnection when a wake word is detected."""
        candidate = _WakeCandidate(connection, score, wake_word)
        self._wake_candidates.append(candidate)
        logger.info(
            "Wake candidate: %s score=%d wake_word=%r (%d in window)",
            connection.name,
            score,
            wake_word,
            len(self._wake_candidates),
        )

        if len(self._wake_candidates) == 1:
            # First candidate — start the arbitration timer.
            loop = asyncio.get_running_loop()
            self._arbitration_handle = loop.call_later(
                ARBITRATION_WINDOW_S,
                lambda: asyncio.ensure_future(self._resolve_arbitration()),
            )

    async def _resolve_arbitration(self) -> None:
        """Pick the best device and start its pipeline; abort the rest."""
        self._arbitration_handle = None
        candidates = self._wake_candidates
        self._wake_candidates = []

        if not candidates:
            return

        # Sort by score descending — highest audio energy wins.
        candidates.sort(key=lambda c: c.score, reverse=True)
        winner = candidates[0]
        losers = candidates[1:]

        logger.info(
            "Wake arbitration resolved: winner=%s (score=%d) out of %d candidate(s)",
            winner.connection.name,
            winner.score,
            len(candidates),
        )

        await winner.connection.start_pipeline(winner.wake_word)

        for loser in losers:
            logger.info(
                "  loser=%s (score=%d)",
                loser.connection.name,
                loser.score,
            )
            await loser.connection.abort_wake()

File: internal/pipeline/device_manager.go
package pipeline

import (
	"context"
	"fmt"
	"github.com/bryfur/ovi-voice-assistant/internal/agent/scheduler"
	"github.com/bryfur/ovi-voice-assistant/internal/device"
	"github.com/bryfur/ovi-voice-assistant/internal/device/codec"
	"log/slog"
	"sort"
	"sync"
	"time"

	"github.com/bryfur/ovi-voice-assistant/internal/config"
	"github.com/bryfur/ovi-voice-assistant/internal/music"
)

// arbitrationWindow is how long to wait for competing wake events before
// picking a winner.
const arbitrationWindow = 500 * time.Millisecond

// wakeCandidate is a device that reported a wake word in the current window.
type wakeCandidate struct {
	conn     *DeviceConnection
	score    int
	wakeWord string
}

// TransportFactory builds a transport for a device; tests may override it.
type TransportFactory func(dev config.DeviceConfig) device.Transport

// DeviceManager creates and manages DeviceConnection instances for WiFi
// devices and arbitrates competing wake words.
type DeviceManager struct {
	connections []*DeviceConnection
	multiDevice bool
	musicGroup  *music.MusicGroup

	mu         sync.Mutex
	candidates []wakeCandidate
	timer      *time.Timer

	// Window is the arbitration window.
	Window time.Duration
}

// NewDeviceManager builds connections for the configured devices.
func NewDeviceManager(devices []config.DeviceConfig, settings *config.Settings, p Runner, ttsRate int, factory TransportFactory) (*DeviceManager, error) {
	if factory == nil {
		factory = func(dev config.DeviceConfig) device.Transport {
			return device.NewWiFiTransport(dev.Host, dev.Port, dev.EncryptionKey)
		}
	}
	m := &DeviceManager{multiDevice: len(devices) > 1, Window: arbitrationWindow}

	// Shared music group for synchronized multi-device playback. Created
	// here so all devices share the same player/queue.
	musicCodec, err := codec.NewCodec(settings.Transport.Codec, 48000, 2, codec.LC3MusicNByte)
	if err != nil {
		return nil, fmt.Errorf("music codec: %w", err)
	}
	m.musicGroup = music.NewMusicGroup(musicCodec.SampleRate(), musicCodec.Channels(), music.Browsers())

	// Only use the arbitration callback when there are multiple devices.
	var onWake WakeCallback
	if m.multiDevice {
		onWake = m.onWake
	}
	for _, dev := range devices {
		c, err := codec.NewCodec(settings.Transport.Codec, ttsRate, 1, 0)
		if err != nil {
			return nil, err
		}
		m.connections = append(m.connections, NewDeviceConnection(
			factory(dev), c, p, settings,
			DeviceOptions{OnWake: onWake, Name: dev.Host, MusicGroup: m.musicGroup},
		))
	}
	return m, nil
}

// Connections returns the managed connections.
func (m *DeviceManager) Connections() []*DeviceConnection { return m.connections }

// MultiDevice reports whether arbitration is enabled.
func (m *DeviceManager) MultiDevice() bool { return m.multiDevice }

// Start connects all devices.
func (m *DeviceManager) Start() error {
	for _, c := range m.connections {
		if err := c.Start(); err != nil {
			return fmt.Errorf("device %s: %w", c.Name, err)
		}
	}
	if m.multiDevice {
		slog.Info("Managing devices — multi-device arbitration enabled",
			"count", len(m.connections), "window", m.Window)
	} else {
		slog.Info("Managing 1 device")
	}
	return nil
}

// Stop disconnects all devices.
func (m *DeviceManager) Stop(ctx context.Context) {
	m.mu.Lock()
	if m.timer != nil {
		m.timer.Stop()
		m.timer = nil
	}
	m.mu.Unlock()
	m.musicGroup.Close(ctx)
	for _, c := range m.connections {
		if err := c.Stop(); err != nil {
			slog.Debug("Device stop error", "device", c.Name, "err", err)
		}
	}
}

// AnnounceAll announces text on every connected device.
func (m *DeviceManager) AnnounceAll(ctx context.Context, text string) error {
	for _, c := range m.connections {
		c.Announce(text)
	}
	return nil
}

// SetScheduler attaches the scheduler to all device connections.
func (m *DeviceManager) SetScheduler(s *scheduler.Scheduler) {
	for _, c := range m.connections {
		c.SetScheduler(s)
	}
}

// -- Wake-word arbitration --

// onWake is called by any DeviceConnection when a wake word is detected.
func (m *DeviceManager) onWake(conn *DeviceConnection, score int, wakeWord string) {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.candidates = append(m.candidates, wakeCandidate{conn, score, wakeWord})
	slog.Info("Wake candidate", "device", conn.Name, "score", score, "wake_word", wakeWord, "in_window", len(m.candidates))
	if len(m.candidates) == 1 {
		// First candidate — start the arbitration timer.
		m.timer = time.AfterFunc(m.Window, m.ResolveArbitration)
	}
}

// PendingCandidates returns the number of candidates in the current window.
func (m *DeviceManager) PendingCandidates() int {
	m.mu.Lock()
	defer m.mu.Unlock()
	return len(m.candidates)
}

// ResolveArbitration picks the best device and starts its pipeline; the
// rest are aborted.
func (m *DeviceManager) ResolveArbitration() {
	m.mu.Lock()
	m.timer = nil
	candidates := m.candidates
	m.candidates = nil
	m.mu.Unlock()
	if len(candidates) == 0 {
		return
	}
	// Sort by score descending — highest normalised energy wins.
	sort.SliceStable(candidates, func(i, j int) bool { return candidates[i].score > candidates[j].score })
	winner := candidates[0]
	slog.Info("Wake arbitration resolved", "winner", winner.conn.Name, "score", winner.score, "candidates", len(candidates))
	winner.conn.StartPipeline(winner.wakeWord)
	for _, loser := range candidates[1:] {
		slog.Info("  loser", "device", loser.conn.Name, "score", loser.score)
		loser.conn.AbortWake()
	}
}
